### PR TITLE
Refactor Spec2Vec and allow both Spectrum and SpectrumDocument input

### PR DESCRIPTION
**Pull request by florian-huber**
Address #51 to make the API easier for users, and prevent issues with different decimal rounding.

- `Spec2Vec` similarity methods now work with `Spectrum` and `SpectrumDocument` input
- code examples were updated (and fixed)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -7,6 +7,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
+## Changed
+
+- refactored `Spec2Vec` to now accept both `Spectrum` and `SpectrumDocument` input [#51](https://github.com/iomega/spec2vec/issues/51)
```
Comment (sverhoeven) [inline CHANGELOG.md]: Sounds like you need to give Spec2Vec both Spectrum and SpectrumDocument at the same time while either one is required.
```suggestion
- refactored `Spec2Vec` to now accept `Spectrum` or `SpectrumDocument` as input [#51](https://github.com/iomega/spec2vec/issues/51)
```

File: spec2vec/Spec2Vec.py [hunk 2]
```diff
@@ -69,43 +100,44 @@ def __init__(self, model: Word2Vec, intensity_weighting_power: Union[float, int]
...
-                                       self.allowed_missing_percentage)
-        query_vector = calc_vector(self.model, query, self.intensity_weighting_power,
-                                   self.allowed_missing_percentage)
+        reference_vector = self._calculate_embedding(reference)
+        query_vector = self._calculate_embedding(query)
 
         return cosine_similarity(reference_vector, query_vector)
 
-    def matrix(self, references: List[SpectrumDocument], queries: List[SpectrumDocument],
+    def matrix(self, references: Union[List[SpectrumDocument], List[SpectrumType]],
```
Comment (sverhoeven) [inline spec2vec/Spec2Vec.py]: Probably `List[Union[SpectrumDocument, SpectrumType]]`could also work, but the type now is easier to work with as  all list members are expected to have the same type. So what you have is more future proof.

File: spec2vec/Spec2Vec.py [hunk 3]
```diff
@@ -69,43 +100,44 @@ def __init__(self, model: Word2Vec, intensity_weighting_power: Union[float, int]
...
+               queries: Union[List[SpectrumDocument], List[SpectrumType]],
                is_symmetric: bool = False) -> numpy.ndarray:
         """Calculate the spec2vec similarities between all references and queries.
 
         Parameters
         ----------
         references:
-            Reference spectrum documents.
+            Reference spectrums or spectrum documents
```
Comment (sverhoeven) [inline spec2vec/Spec2Vec.py]: Use consistent punctuation
```suggestion
            Reference spectrums or spectrum documents.
```

File: spec2vec/Spec2Vec.py [hunk 4]
```diff
@@ -132,11 +162,26 @@ def matrix(self, references: List[SpectrumDocument], queries: List[SpectrumDocum
...
+        word_regex = r"[a-z]{4}@[0-9]{1,5}."
+        example_word = next(iter(model.wv.vocab))
+        return len(re.split(word_regex, example_word)[-1])
+
+    def _calculate_embedding(self, spectrum_in: Union[SpectrumDocument, SpectrumType]):
+        """Generate Spec2Vec embedding vectors from input spectrum (or SpectrumDocument)"""
+        if isinstance(spectrum_in, Spectrum):
+            spectrum_in = SpectrumDocument(spectrum_in, n_decimals=self.n_decimals)
+        assert spectrum_in.n_decimals == self.n_decimals, \
+            "Decimal rounding of input data does not agree with model vocabulary."
```
Comment (sverhoeven) [inline spec2vec/Spec2Vec.py]: This can only be False when spectrum_in is already a SpectrumDocument. So assert could be skipped when spectrum_in is a Spectrum
```suggestion
        else:
            assert spectrum_in.n_decimals == self.n_decimals, \
                "Decimal rounding of input data does not agree with model vocabulary."
```

File: spec2vec/Spec2Vec.py [hunk 5]
```diff
@@ -20,31 +23,59 @@ class Spec2Vec(BaseSimilarity):
...
+            s = normalize_intensities(s)
+            s = select_by_mz(s, mz_from=0, mz_to=1000)
+            s = select_by_intensity(s, intensity_from=0.01)
+            s = add_losses(s, loss_mz_from=10.0, loss_mz_to=200.0)
+            s = require_minimum_number_of_peaks(s, n_required=5)
+            return s
 
-        # Import pre-trained word2vec model (alternative: train new model)
-        model_file = "path and filename"
+        spectrums_file = os.path.join(os.getcwd(), "..", "tests", "pesticides.mgf")
```
Comment (sverhoeven) [inline spec2vec/Spec2Vec.py:53]: It will be hard to find this file reading this on https://spec2vec.readthedocs.io/en/latest/. Can you add a download link to the example intro text with 
```rst
:download:`../tests/pesticides.mgf </../tests/pesticides.mgf>`
:download:`../integration-tests/test_user_workflow_spec2vec.model </../integration-tests/test_user_workflow_spec2vec.model>`
```
So people dont have to go GitHub to find the file.

File: spec2vec/Spec2Vec.py [hunk 6]
```diff
@@ -20,31 +23,59 @@ class Spec2Vec(BaseSimilarity):
...
+        # Select top-10 candidates for first query spectrum
+        spectrum0_top10 = scores.scores_by_query(spectrums[0], sort=True)[:10]
 
-        # Filter out self-comparisons
-        filtered = [(reference, query, score) for (reference, query, score) in scores if reference != query]
+    Should output
+
+    .. testoutput::
+
+        Removed adduct M-H from compound name.
```
Comment (sverhoeven) [inline spec2vec/Spec2Vec.py:81]: The test output is a bit minimal. Could we print the spectrum0_top10? Something like

```python
>>> [s[0].metadata['spectrumid'] for s in spectrum0_top10]
['CCMSLIB00001058300',
 'CCMSLIB00001058289',
 'CCMSLIB00001058303',
 'CCMSLIB00001058433',
 'CCMSLIB00001058430',
 'CCMSLIB00001058313',
 'CCMSLIB00001058426',
 'CCMSLIB00001058394',
 'CCMSLIB00001058298',
 'CCMSLIB00001058317']

```

```

**Review by sverhoeven — CHANGES_REQUESTED** [review]
Nice to see improved compatibility with matchms. You retained the high code coverage and documentation renders nicely.

SpectrumType is defined as Optional[Spectrum], so it can be None, how does your code deal with None cases? Can you add tests for when spectrum = None?

I got `AssertionError: Could not find known_adduct_conversions.csv.
` error when I tried to run the Spec2Vec example code because the matchms 0.8.0 from https://anaconda.org/bioconda/matchms/ is missing the data files. It seems we have matchms on bioconda and nlesc channel. And when I did `conda env create --file conda/environment-dev.yml` the matchms from bioconda was installed. After I uninstalled/installed with pip I could run the example.

```
conda list |grep matchms
matchms                   0.8.0              pyh3252c3a_0    bioconda
```

File: spec2vec/Spec2Vec.py (hunk 2 above)
Comment (florian-huber) [reply spec2vec/Spec2Vec.py]: I also briefly thought about that. But I was afraid that `List[Union[A, B]]` would suggest that could can also have a list with mixed types. In the current implementation that would even work, but it doesn't seem recommended use to me .

File: spec2vec/Spec2Vec.py (hunk 4 above)
Comment (florian-huber) [reply spec2vec/Spec2Vec.py]: Good point. Thanks.

File: spec2vec/Spec2Vec.py (hunk 6 above)
Comment (florian-huber) [reply spec2vec/Spec2Vec.py:81]: That's much nicer indeed. Added this.

File: spec2vec/Spec2Vec.py (hunk 5 above)
Comment (florian-huber) [reply spec2vec/Spec2Vec.py:53]: I will try adding this.

**Comment by florian-huber** [comment]
> SpectrumType is defined as Optional[Spectrum], so it can be None, how does your code deal with None cases? Can you add tests for when spectrum = None?

Thanks for spotting this Stefan! I think `None` as input should not be allowed, so I changed the input type from `SpectrumType` to `Spectrum` and added that it will raise an `ValueError` if the input type is not `Spectrum` or `SpectrumDocument`.

**Comment by florian-huber** [comment]
> I got AssertionError: Could not find known_adduct_conversions.csv. error when I tried to run the Spec2Vec example code because ...

I added the package_data to the pypi `.tar.gz` and `matchms 0.8.1` now works (the bioconda part still misses rdkit, but that should better be dealt with in https://github.com/matchms/matchms/issues/178)

File: spec2vec/Spec2Vec.py (hunk 5 above)
Comment (florian-huber) [reply spec2vec/Spec2Vec.py:53]: Added this to the docstring. But then also realized I have no admin rights on readthedocs so I couldn't inspect the results.

Comment (sverhoeven) [reply spec2vec/Spec2Vec.py:53]: You can build the docs locally with `cd readthedocs;make html;google-chrome _build/html/index.html`.

Added you as maintainer of https://readthedocs.org/dashboard/spec2vec/users/

**Review by sverhoeven — APPROVED** [review]
Thanks for the None exception+test. The None makes sense when calling filtering functions, but not for calculating scores.

File: spec2vec/Spec2Vec.py (hunk 5 above)
Comment (florian-huber) [reply spec2vec/Spec2Vec.py:53]: Perfect, thanks Stefan.
Don't know why I didn't think of rendering it locally.... but that of course works nicely :)
(and thanks for adding me as admin).

**Comment by florian-huber** [comment]
Thanks a lot for the review Stefan!
